### DISCUSSION TITLE
Support defining data schema for storage

### DISCUSSION THREAD

**Discussion by jalooc** [wxt-dev/wxt#1123] [Ideas]
While the ability to define the storage data type when using functions like `storage.defineItem<Type>()` is convenient, it's sadly not really type-safe - the contents of the storage can be modified either by code not using the same storage util, or even manually, so it's not guaranteed that we always get data of type `Type` when using `getValue()`.

The idea is to allow passing a schema (preferably a [zod](https://github.com/colinhacks/zod) schema) against which the value read from the storage would be validated internally in `getValue()` before being returned. A nice bonus is that it would be easy to automatically infer the type based on the schema, so the dev wouldn't have to specify it on top of the schema.

It could look sth like this:
```ts
const myStorage = storage.defineItem('local:my-storage', z.object({ foo: z.number() }))

myStorage.getValue().then(value => /* fully type-safe `{ foo: number }` */)
```
and the types for `defineItem()` would go sth like this:
```ts
import { z } from 'zod'

const defineItem = <Schema extends z.ZodType<unknown>>(key: string, schema: Schema) => {
  getValue: () => z.infer<Schema>,
}
```

Alternatively, if we wanted to avoid vendor-lockin on zod, we could just have a `validate` parameter:
```ts
const defineItem = <T>(key: string, validate: () => T) => {
  getValue: () => T,
}
```

**Comment by aklinker1** [comment]
This would be a good thing to add, along with serialize/deserialize options. Created an issue to track it: https://github.com/wxt-dev/wxt/issues/1173
